pkg/sampling/w3ctracestate.go

```diff
@@ -0,0 +1,165 @@
+	//
+	// Note the use of double-quoted strings in two places below.
+	// This is for \x expansion in these two cases.  Also note
+	// \x2d is a hyphen character, so a quoted \ (i.e., \\\x2d)
+	// appears below.
+	valueNonblankCharRegexp = "[\x21-\x2b\\\x2d-\x3c\x3e-\x7e]"
+	valueCharRegexp         = "[\x20-\x2b\\\x2d-\x3c\x3e-\x7e]"
+	valueRegexp             = valueCharRegexp + `{0,255}` + valueNonblankCharRegexp
+
+	// tracestate  = list-member 0*31( OWS "," OWS list-member )
+	// list-member = (key "=" value) / OWS
+
+	owsCharSet      = ` \t`
+	owsRegexp       = `(?:[` + owsCharSet + `]*)`
+	w3cMemberRegexp = `(?:` + keyRegexp + `=` + valueRegexp + `)?`
+
+	// This regexp is large enough that regexp impl refuses to
+	// make 31 copies of it (i.e., `{0,31}`) so we use `*` below.
+	w3cOwsMemberOwsRegexp      = `(?:` + owsRegexp + w3cMemberRegexp + owsRegexp + `)`
+	w3cCommaOwsMemberOwsRegexp = `(?:` + `,` + w3cOwsMemberOwsRegexp + `)`
+
+	// The limit to 31 of owsCommaMemberRegexp is applied in code.
+	w3cTracestateRegexp = `^` + w3cOwsMemberOwsRegexp + w3cCommaOwsMemberOwsRegexp + `*$`
+)
+
+var (
+	w3cTracestateRe = regexp.MustCompile(w3cTracestateRegexp)
+
+	w3cSyntax = keyValueScanner{
+		maxItems:  hardMaxNumPairs,
+		trim:      true,
+		separator: ',',
+		equality:  '=',
+	}
+)
+
+// NewW3CTraceState parses a W3C trace state, with special attention
+// to the embedded OpenTelemetry trace state field.
+func NewW3CTraceState(input string) (w3c W3CTraceState, _ error) {
+	if len(input) > hardMaxW3CLength {
+		return w3c, ErrTraceStateSize
+	}
+
+	if !w3cTracestateRe.MatchString(input) {
+		return w3c, strconv.ErrSyntax
+	}
+
+	err := w3cSyntax.scanKeyValues(input, func(key, value string) error {
+		if len(key) > hardMaxKeyLength {
+			return ErrTraceStateSize
+		}
+		if tenant, system, found := strings.Cut(key, multiTenantSep); found {
+			if len(tenant) > hardMaxTenantLength {
+				return ErrTraceStateSize
+			}
+			if len(system) > hardMaxSystemLength {
+				return ErrTraceStateSize
+			}
+		}
+		switch key {
+		case otelVendorCode:
+			var err error
+			w3c.otts, err = NewOTelTraceState(value)
+			return err
+		default:
+			w3c.kvs = append(w3c.kvs, KV{
+				Key:   key,
+				Value: value,
+			})
+			return nil
+		}
+	})
+	return w3c, err
+}
+
+// HasAnyValue indicates whether there are any values in this
+// tracestate, including extra values.
+func (w3c *W3CTraceState) HasAnyValue() bool {
+	return w3c.HasOTelValue() || w3c.HasExtraValues()
+}
+
+// OTelValue returns the OpenTelemetry tracestate value.
+func (w3c *W3CTraceState) OTelValue() *OTelTraceState {
+	return &w3c.otts
+}
+
+// HasOTelValue indicates whether an OpenTelemetry tracestate value
+// is present in this tracestate.
+func (w3c *W3CTraceState) HasOTelValue() bool {
+	return w3c.otts.HasAnyValue()
+}
+
+// Serialize encodes this tracestate object for use as a W3C
+// tracestate header value.
+func (w3c *W3CTraceState) Serialize(w io.StringWriter) error {
```

Comment:
I take it the idea is to stop adding fields before they exceed a limit, so that the result is parsable, even if it has lost some fields? I considered whether a user might pass a fixed-size StringWriter and let it fail when there are too many bytes, but there isn't an off-the-shelf method to do that. Willing to do this.

Reply:
I think we can do this later. Given the state of the issues list in this repository, I do not think it is worthwhile to record an issue to track the work. If someone complains about wanting certain truncation behavior because a vendor is rejecting span data, that'll help motivate us.

Also, this topic will come up in the review for my next PR unavoidably. If the tracestate can't be parsed, how should a sampler treat that? I think we'll probably want to let the user configure one of several choices, such as:
- let the span pass
- drop the span
- return an error

Reply:
I'm happy leaving it for later -- IIRC, the spec calls for writing new fields at the front of the string, and dropping old fields that don't fit at the end of the string (never truncating in a way that breaks fields).